File: vehicles.c
#include <stdio.h>

#define NUM_VEHICLES 5

struct Vehicle {
    char id[20];
    char type[30];
    double price;
};

int main(void) {
    struct Vehicle vehicles[NUM_VEHICLES];

    for (int i = 0; i < NUM_VEHICLES; i++) {
        printf("Vehicle %d\n", i + 1);

        printf("  Vehicle ID: ");
        if (scanf("%19s", vehicles[i].id) != 1) {
            fprintf(stderr, "Failed to read vehicle ID.\n");
            return 1;
        }

        printf("  Type: ");
        if (scanf("%29s", vehicles[i].type) != 1) {
            fprintf(stderr, "Failed to read type.\n");
            return 1;
        }

        printf("  Price: ");
        if (scanf("%lf", &vehicles[i].price) != 1) {
            fprintf(stderr, "Failed to read price.\n");
            return 1;
        }
    }

    int most = 0;
    int least = 0;
    for (int i = 1; i < NUM_VEHICLES; i++) {
        if (vehicles[i].price > vehicles[most].price) {
            most = i;
        }
        if (vehicles[i].price < vehicles[least].price) {
            least = i;
        }
    }

    printf("\nMost expensive vehicle:\n");
    printf("  ID: %s\n  Type: %s\n  Price: %.2f\n",
           vehicles[most].id, vehicles[most].type, vehicles[most].price);

    printf("\nLeast expensive vehicle:\n");
    printf("  ID: %s\n  Type: %s\n  Price: %.2f\n",
           vehicles[least].id, vehicles[least].type, vehicles[least].price);

    return 0;
}

File: vehicles.js
const NUM_VEHICLES = 5;

const FIELDS = [
  { key: "id", label: "Vehicle ID", type: "text" },
  { key: "type", label: "Type", type: "text" },
  { key: "price", label: "Price", type: "text" },
];

function buildForm(container) {
  for (let i = 0; i < NUM_VEHICLES; i++) {
    const row = document.createElement("fieldset");
    row.className = "vehicle-row";

    FIELDS.forEach((field) => {
      const wrapper = document.createElement("div");
      wrapper.className = "field";

      const input = document.createElement("input");
      input.type = field.type;
      input.id = `${field.key}-${i}`;
      input.name = `${field.key}-${i}`;
      input.dataset.field = field.key;
      input.dataset.index = String(i);

      const label = document.createElement("label");
      label.htmlFor = input.id;
      label.textContent = `Vehicle ${i + 1} — ${field.label}`;

      const error = document.createElement("div");
      error.className = "error";
      error.id = `error-${field.key}-${i}`;

      wrapper.append(label, input, error);
      row.appendChild(wrapper);
    });

    container.appendChild(row);
  }
}

function setError(input, message) {
  const error = document.getElementById(
    `error-${input.dataset.field}-${input.dataset.index}`
  );
  error.textContent = message;
  input.classList.toggle("invalid", Boolean(message));
  input.setAttribute("aria-invalid", message ? "true" : "false");
}

function validateInput(input) {
  const value = input.value.trim();
  const index = Number(input.dataset.index) + 1;

  if (value === "") {
    setError(input, `Vehicle ${index}: this field is required.`);
    return null;
  }

  if (input.dataset.field === "price") {
    const price = Number(value);
    if (!Number.isFinite(price)) {
      setError(input, `Vehicle ${index}: price must be a number.`);
      return null;
    }
    if (price <= 0) {
      setError(input, `Vehicle ${index}: price must be greater than 0.`);
      return null;
    }
    setError(input, "");
    return price;
  }

  setError(input, "");
  return value;
}

function collectVehicles(form) {
  const vehicles = [];
  let valid = true;

  for (let i = 0; i < NUM_VEHICLES; i++) {
    const vehicle = {};
    FIELDS.forEach((field) => {
      const input = form.querySelector(`#${field.key}-${i}`);
      const value = validateInput(input);
      if (value === null) {
        valid = false;
      } else {
        vehicle[field.key] = value;
      }
    });
    vehicles.push(vehicle);
  }

  return valid ? vehicles : null;
}

function formatPrice(price) {
  return price.toLocaleString(undefined, {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}

function renderResult(section, vehicles) {
  let most = vehicles[0];
  let least = vehicles[0];

  vehicles.forEach((vehicle) => {
    if (vehicle.price > most.price) most = vehicle;
    if (vehicle.price < least.price) least = vehicle;
  });

  const heading = document.createElement("h2");
  heading.textContent = "Results";

  const line = (title, vehicle) => {
    const paragraph = document.createElement("p");
    const label = document.createElement("strong");
    label.textContent = `${title}: `;
    paragraph.append(
      label,
      `${vehicle.id} (${vehicle.type}) — ${formatPrice(vehicle.price)}`
    );
    return paragraph;
  };

  section.replaceChildren(
    heading,
    line("Most expensive", most),
    line("Least expensive", least)
  );
  section.hidden = false;
}

document.addEventListener("DOMContentLoaded", () => {
  const form = document.getElementById("vehicle-form");
  const container = document.getElementById("vehicle-fields");
  const result = document.getElementById("result");

  buildForm(container);

  container.addEventListener("input", (event) => {
    if (event.target instanceof HTMLInputElement) {
      validateInput(event.target);
    }
  });

  form.addEventListener("submit", (event) => {
    event.preventDefault();
    const vehicles = collectVehicles(form);

    if (!vehicles) {
      result.hidden = true;
      const firstInvalid = form.querySelector("input.invalid");
      if (firstInvalid) firstInvalid.focus();
      return;
    }

    renderResult(result, vehicles);
  });
});
